pyacq/core/nodegroup.py

```diff
@@ -1,93 +1,161 @@
 from pyqtgraph.Qt import QtCore, QtGui
+from pyqtgraph.util.mutex import Mutex
 
 from .rpc import RPCServer
-from .node import Node, WidgetNode
+from .node import Node, WidgetNode, register_node_type
 from .nodelist import all_nodes
 
+import time
+import importlib
+import pickle
+import weakref
+import random
+import string
+import zmq
 
-
-class RpcThread( QtCore.QThread):
-    def __init__(self, rpc_server, parent = None):
+class RpcThreadSocket( QtCore.QThread):
+    new_message = QtCore.Signal(QtCore.QByteArray, QtCore.QByteArray)
+    def __init__(self, rpc_server, local_addr, parent = None):
         QtCore.QThread.__init__(self, parent)
-        self.rpc_server = rpc_server
+        self.rpc_socket = weakref.ref(rpc_server._socket)
+        
+        context = zmq.Context.instance()
+        self.local_socket = context.socket(zmq.PAIR)
```

Comment:
You say this for local_socket or rpc_socket ?
I have read in zmq that it is sane to instanciate socket in a thread and use them in another thread when there a guarantee that the socket is not manipulated in the 2 places.
In http://zguide.zeromq.org/page:all, they said
"Technically it's possible to migrate a socket from one thread to another but it demands skill."